internal/service/advancedclustertpf/resource.go

```
@@ -386,83 +383,50 @@ func (r *rs) applyAdvancedConfigurationChanges(ctx context.Context, diags *diag.
 }
 
 func (r *rs) applyClusterChanges(ctx context.Context, diags *diag.Diagnostics, state, plan *TFModel, patchReq *admin.ClusterDescription20240805) *admin.ClusterDescription20240805 {
-	var cluster *admin.ClusterDescription20240805
-	if usingLegacySchema(ctx, plan.ReplicationSpecs, diags) {
-		// Only updates of replication specs will be done with legacy API
-		legacySpecsChanged := r.updateLegacyReplicationSpecs(ctx, state, plan, diags, patchReq.ReplicationSpecs)
+	if usingLegacyShardingConfig(ctx, plan.ReplicationSpecs, diags) {
+		// With old sharding config we call older API (2023-02-01) for updating replication specs to avoid cluster having asymmetric autoscaling mode. Old sharding config can only represent symmetric clusters.
+		r.updateLegacyReplicationSpecs(ctx, state, plan, diags, patchReq.ReplicationSpecs)
 		if diags.HasError() {
 			return nil
 		}
-		patchReq.ReplicationSpecs = nil // Already updated by legacy API
-		if legacySpecsChanged && update.IsZeroValues(patchReq) {
+		patchReq.ReplicationSpecs = nil // Already updated by 2023-02-01 API
+		if update.IsZeroValues(patchReq) {
 			return AwaitChanges(ctx, r.Client.AtlasV2.ClustersApi, &plan.Timeouts, diags, plan.ProjectID.ValueString(), plan.Name.ValueString(), changeReasonUpdate)
 		}
 	}
-	if update.IsZeroValues(patchReq) {
-		return cluster
-	}
-	pauseAfter := false
-	if patchReq.Paused != nil && patchReq.GetPaused() {
-		// More changes than pause, need to pause after
-		if !reflect.DeepEqual(pauseRequest, *patchReq) {
-			pauseAfter = true
-			patchReq.Paused = nil
-		}
-	} else if patchReq.Paused != nil && !patchReq.GetPaused() {
-		// More changes than pause, need to resume before applying changes
-		if !reflect.DeepEqual(resumeRequest, *patchReq) {
-			patchReq.Paused = nil
-			_ = r.updateAndWait(ctx, &resumeRequest, diags, plan)
-		}
-	}
```

Not finding a reason of handling the paused change in a separate request, and can see it was leading to more complexity.
For the case of `More changes than pause, need to resume before applying changes` we are clear in our docs that paused cluster do not support updates ([reference](https://registry.terraform.io/providers/mongodb/mongodbatlas/latest/docs/resources/advanced_cluster#paused-1)).

---

nice simplification.  mentioning @EspenAlbert in case he sees some problem

---

I thought this was an `artifical` limitation in our old implementation.
Removing this logic will not make it possible to resume a cluster that has been paused and accidentally has some extra changes.
But I really like the simplification

---

Correct, I would consider that use case 1 is a current limitation in SDK v2 implementation as well. For case 2 we would be sending sending all request info in a single PATCH (paused `true` + additional changes) for the case of new sharding config, so if API does not support this I believe we can simply return that same error. Again as you mention if paused change is performed in isolated manner there is no risk of errors.

---

what happens in step 1, we'll send an error or cluster could get in a inconsistent state?

@AgustinBettati @EspenAlbert just to double check, also cc @marcosuma , as it looks like we have the opportunity to allow users to make pause with other changes at the same time (managing this in TF side even if Atlas doesn't support it),  do we want to allow that capability, or we prefer to keep the SDKv2 limitation and don't allow extra changes when pause is updated?

---

My feeling is that we are removing business logic and adhering to the the API behaviour. This would force users to two operations what they were doing with one. Am I right? What is the strong reason we are making this change? Is it about simplifying our code base?

---

@marcosuma No, right now they need  2 operations,  the improve in TPF was to allow both changes in 1 operation. 
In general we're maintaining parity with SDKv2 and keep improvements for later, as in other cases like failing if cluster name changes instead of deleting and creating a new one.

I suppose we want to follow the same tenant here and simplify the code, and have the user to need 2 operations in SDKv2.   

but as the code was already done, we might want to keep the functionality, but we can also do it in the future.


So I think that I probably agree with @AgustinBettati to keep SDKv2 behavior and code simplification, and leave this improvement for the future cc @EspenAlbert

---

Sorry, had overlooked some details in SDK v2 implementation which do simplify one of the use cases.

**Resume a paused cluster and also add additional changes** (no difference):
- SDK v2: Done in same PATCH request, this returns error `Cannot update cluster while it is paused or being paused.`
- TPF current PR: Both paused and other changes in same PATCH request, would have same error as SDK v2.

**Pause a cluster and also add additional changes** (difference):
- SDK v2: Sends pausing request as a follow up request after doing other changes. Avoids error `Cannot update and pause cluster at the same time.`.
- TPF current PR: Both paused and other changes in same PATCH request, has error `Cannot update and pause cluster at the same time.`.

Will follow up with change/revert in TPF so `Pause a cluster and also add additional changes` is supported as in SDK v2 and we ensure no regression. If `Resume a paused cluster and also add additional changes` is also simple will factor in as well (however this not supported in SDK v2).

---

@AgustinBettati  I'd prioritize for now the "feature parity" so based on your last comment it should really be about making "Pause a cluster and also add additional changes" behave the same

---

https://github.com/mongodb/terraform-provider-mongodbatlas/pull/2973/commits/94c90c01f639301491fa18841d93441d401f296b adds capability of `Pause a cluster and also add additional changes`, also modified the existing acceptance test (that runs in SDK v2 and TPF) to capture this. `Resume a paused cluster and also add additional changes` was trivial to include so added as well.

---

Nice @AgustinBettati LGTM